src/bin/pgcopydb/ld_replay.c

```diff
@@ -249,3 +255,166 @@ stream_replay_line(void *ctx, const char *line, bool *stop)
 
 	return true;
 }
+
+
+/*
+ * streamBufferedXID reads a transaction buffer file and streams it into the
+ * target database.
+ */
+static bool
+streamBufferedXID(ReplayStreamCtx *replayCtx)
+{
+	/* Open replaybuffer in read-only mode to use it as input stream */
+	FILE *txnBufferForRead = fopen_read_only(replayCtx->currentTxnFileName);
+
+	if (txnBufferForRead == NULL)
+	{
+		log_error("Failed to open transaction buffer file \"%s\" in "
+				  "readonly mode: %m",
+				  replayCtx->currentTxnFileName);
+		return false;
+	}
+
+	ReadFromStreamContext readerContext = {
+		.callback = stream_replay_line,
+		.ctx = replayCtx
+	};
+
+	if (!read_from_stream(txnBufferForRead, &readerContext))
+	{
+		log_error("Failed to read from replay buffer, "
+				  "see above for details");
+		return false;
+	}
+
+	fclose(txnBufferForRead);
+
+	return true;
+}
+
+
+/* streamReplayLineIntoBuffer is a callback function for the ReadFromStreamContext
+ * and read_from_stream infrastructure. It's called on each line read from a
+ * stream such as a unix pipe and buffers the line into a file until we encounter
+ * a COMMIT. This ensures that we can replay the whole transaction in one go
+ * and avoids deadlock when the BEGIN messages arrives before the existence of
+ * transation metadata file which is created by the transform process.
+ */
```

Comment:
Is there a way to activate buffering only when a transaction does not have its COMMIT LSN attached in the metadata? So that in the case when we have a typical OTLP system with many one-command DML transactions we don't have any penalty?

Reply:
Thanks!